### ISSUE TITLE
OnDemandGrid, as a *tree grid*, makes too many JsonRest requests when scrolling children

### ISSUE THREAD

**Issue by frankf2** [SitePen/dgrid#877]
**Test Files:**
* Edit dgrid/test/data/rest.php to change $total=500 to **$total=5000**.
* Open dgrid/test/JsonRest.html in a web browser, with the debug console displayed.
* Note:  In rest.php, $total=5000 sets the “total” number of items available for not only the top level, but for any/all expanded item’s children.

**Duplicate the Problem:**
* Without opening a top level item, quickly drag the scrollbar from the top to the bottom.  The console shows that dgrid *only* asks for the last/end Range of the 5000 top level items.
* Open one of the top level items, e.g., the bottom one.
* Quickly drag the scrollbar from the top to the bottom, as before.  The console shows that dgrid requests ALL the items, 250 rows at a time, for the 5000 children of the top level item.

**Net:**  dgrid asks for the correct items (very limited set) when scrolling top level items, but asks for ALL children items when scrolling a top level item’s children items.

**Expected Behavior:**  When scrolling a dgrid, dgrid should make the same limited item request when scrolling an item’s children as it does when scrolling a top level item.

**Comment by frankf2** [comment]
Is there an update on this, hopefully fixed?  We are on 0.4, but are still having problems.  Or is there another grid that would be more suitable?

This is today's comment from one of our developers: 

"This is a problem that Frank reported for us to SitePen back in Mar. 2014. It is still unresolved."

"I thought we’d be able to use dgrid’s scrollTo() function if we could figure out the position to scroll to. But, even if we do that, dgrid will fetch ALL of the records  from zero to the position that we scroll to.  For example, if we tell dgrid to scrollTo a position that is 5000 rows below our current location dgrid makes over 20 calls to the server to fetch all of the records (it asks for 250 rows at a time when we make that type of large scroll request).  That won’t work, it’s too slow. "

**Comment by eQuorumRayNewman** [comment]
This continues to be a major issue for us and is impacting our sales and customer satisfaction.  Is there fix in the works or on the schedule?  If not who can I contact to explore the possibility of paid consulting to get an acceptable solution.

**Comment by dylans** [comment]
Hi @eQuorumRayNewman, it is a challenging issue and we've simply not had time to make it a priority. You can feel free to reach out to me ( dylan at sitepen dot com ) to discuss our options to make this fix a priority.

**Comment by mozg0123** [comment]
Hello, I think I found a solution: it's just a bug that always reproduced with such structure. Fixing the bug allows to make it work in such case properly. Perhaps the fix can cause another bugs or even fix another bugs. But, it's changing only one line of code, I think it's fixable. After my fix I got only about 30 requests to Server. Only 19 of them are to get child rows. Before the fix it will require about 5000 (all children of one row) / 50 (rows per request) = 100.

Please contact me to discuss the fix in details - mozg0123 at gmail dot com. I assume we will be able to resolve the issue. If it will work - we can try to create a pull request to push it here.
![about 30 requests to server](https://cloud.githubusercontent.com/assets/20361230/16708461/750af014-45fd-11e6-83f6-881f5a60c2bb.png)

**Comment by frankf2** [comment]
mozg0123, Thanks, that sounds encouraging.  Could you just post the possible fix and any considerations here in this thread so that all interested parties could test and collaborate on the potential fix?  We'll then be glad to test the fix for our environment and post our results here.

**Comment by kriszyp** [comment]
@frankf2, want to try #1300 and see if that helps? It doesn't eliminate the possibility of dgrid requiring sequential loads, but within uninterrupted child lists, it should help avoid it.

**Comment by dylans** [comment]
@kriszyp fwiw, @edhager has been working on a fix for this for the past few weeks. It's probably worth collaborating with him before we land a fix, as he's close to complete with that.

**Comment by kriszyp** [comment]
@dylans, ok, great. We were experiencing this issue, so I put this together to address the tree we are working with. If you have another solution, or this helps at all, either way sounds good.

The previous code was definitely not properly generating a full/correct linked list of preload nodes. And correctly this may work if large data sets are only at the leaf level. The limitation I didn't attempt to address was doing far off node removal  across children.

**Comment by edhager** [comment]
@kriszyp, if you are interested, my progress so far is here: https://github.com/edhager/dgrid/tree/preload-tree-refactor.  There is a bit of code from your pull request in there.

Just like you, I fixed the management of the preload list.  In addition, I modified the far off removal code to work across all of the preload nodes.  That means the child rows are being pruned like the top level rows.

Here is my todo list:

- While testing with test/Rest.html, expand the first row, scroll to the bottom and then start scrolling to the top. The grid will make a jump to row 2 earlier than expected. I think something is not quite right with the code that is trying to maintain the scroll position as I scroll up.
- The following code that Ken added, which I commented out, needs to be updated to support managing child items in the store: https://github.com/edhager/dgrid/blob/preload-tree-refactor/OnDemandList.js#L528-L546
- Add new tests and run the test suite.

FYI - I modified the manual test page test/Rest.html to spit out some preload data to the console as the rows are added and removed.
